### ISSUE TITLE
anonymous function parameter name lost during binary generation

### ISSUE THREAD

**Issue by davidfestal** [ceylon/ceylon-compiler#1991]
With such a code :
````ceylon
value simplifier = Match {
    Rule {
        when(And t) => t.right == myTrue;
        do(And t) => t.left;
    }
}
```
when stepping inside the `do()` anonymous function during a debug session,
the `t` parameter is known in JDI with the name `x0`.

In some other cases, the `t` parameter name is kept.

This prevents debug variable hovers, because the Ceylon parameter name and the corresponding JDI local variable name should match.

**Comment by lucaswerkmeister** [comment]
I haven’t been able to find stand-alone code (independent of @davidfestal’s [Travelon](https://github.com/davidfestal/travelon/blob/f439c986259a25fa362f72165c233b737d81b80f/source/com/serli/travelon/run.ceylon#L81)) that produces an `x0` parameter name, but in
```ceylon
class Doer(Anything do(Anything t)) {}
Doer id = Doer {
    do(Anything t) => t;
};
```
the `t` parameter becomes a `$param$0`, which is probably almost as bad for debugging (though at least the `$` makes it clear that it can’t be a “real” Ceylon name).

[generated code](https://gist.github.com/lucaswerkmeister/71534c5ac55b437a3d88)

**Comment by davidfestal** [comment]
@lucaswerkmeister : After filtering out the call to an additional synthetic `$call$` method, I now suspend in a location where the behavior is exactly the same as you describe.

But what is strange is that, if we really stop on the `return` statement, we should see the `t` local variable, and it's not the case :  the location in which JDI stops us when adding a breakpoint inside `do` (even in the block form) doesn't know of the `t` variable.

So if change the code to this :
```Ceylon
Doer id = Doer {
    function do(Anything t) {
       return t;
   }
};
```
The generated code is the same and the problem is the same (`t` not visible).

However if I add a `print(t)`statement just before the `return` statement, it's a bit different :
*__the debugger suspends twice__* on this line :
- the first time with the `t` variable *unavailable*
- and the second time with the `t` variable *visible*

This can be explained by the [generated source](https://gist.github.com/davidfestal/f54616ba4b026063ed6a) containg stored lines added by the `com.redhat.ceylon.compiler.java.codegen.JavaPositionsRetriever` utility.

Mainly by these 3 lines in the body of the `$call$` method :
```Java
                            final .java.lang.Object t = $param$0; // line 5
                            .ceylon.language.print_.print(t); // line 4
                            return t; // line 5
```

The first and last statements are linked to the same line number - 5 - , so that when a breakpoint is added on line 5, we stop twice and the first time can obviously not see the `t` variable since it's not already defined.

When we don't have the `print()` statement, since the 2 statements with the same attached line (5) are adjacent, the program is suspended only once for both lines, and it seems it suspends on the first one, so that the `t` variable is not visible.

In fact, when running such a code ([here](https://gist.github.com/davidfestal/f54616ba4b026063ed6a)) , it can happen that we also suspend at the instanciation of the callable object, which is consistent with the fact that the `new` statement [here](https://gist.github.com/davidfestal/f54616ba4b026063ed6a#file-doer-java-L26) is also associated with line 5, which seems wrong.

@FroMage , @quintesse , @tombentley : it seems to me that having several statements (corresponding to different code indexes in the bytecode) with the same associated Ceylon line number is a generally a bad thing. It seems to me that we should *__only__* add line information to statements that correspond explicitly to a Ceylon statement. And if several generated statements are related to a single Ceylon statement, we should add only the best fit, arguably the one that doesn't lead to information lost (as happened in this issue).

For a correct breakpoint behavior, the real need is to have *__all statement lines of the Ceylon code__* associated to *__only one (the most representative) statement of the generated Java code__*.  

And if all other statements have no line information, it doesn't seem to be a problem IMO. We can even add a step and breakpoint filter that skips suspend requests in such locations.
Moreover, an "unknown line number" problem is easier to identify and report during debugging than wrongly located lines.

So... should I open an issue about that in ceylon-compiler ?

**Comment by davidfestal** [comment]
And for the case when a `$call$` function just delegates to another one, adding line information only on statements that are part the real Ceylon anonymous  function body could be a good solution to automatically skip suspend requests that might occur in delegating functions.

**Comment by tombentley** [comment]
I certainly think we should generate the assignment

    final .java.lang.Object t = $param$0;

without line number information.

In general, the more we remove line number information the less information we have when a compiler bug affects a piece of code. It's annoying enough when there is a compiler bug, but currently you usually get the line number so know where to start looking for a work around while the bug gets fixed. Without line number info you might only know the compilation unit that's afflicted by the compiler bug. I fear that would inevitably mean fewer and/or worse bug reports and longer bug fixes. So removing line number information is not an unmitigated good. However, I don't see how we could simultaneously have it for bug reporting, but not have it for debugging purposes.

**Comment by quintesse** [comment]
I think that the idea of associating a Ceylon line with at most one statement in the generated Java code just isn't feasible , code is generated by knitting together many pieces and in often we won't be able to tell if that piece is the first let alone the most representative.

And if we remove line number information we create the possibility that if for some reason the debugger stops there we won't be able to tell the user to which part of the Ceylon code it is associated. THat doesn't seem like a good idea.

I just think we have to treat these situations like any other bug: we generate erroneous line information (you're above example obviously shouldn't be generating lines 5, 4, 5), so we have to fix it.

What we can maybe do to help this along is to use that visitor of yours to check for obviously or possibly wrong things like lines going backwards or lines being associated with multiple Ceylon lines.

**Comment by davidfestal** [comment]
Thanks for these comments.

@tombentley : I understand the need of having a precise stackTrace in order to better catch the Ceylon context of a bug.
However, even with missing line numbers, the stackTrace could give interesting Ceylon context information with the help of a simple Java-to-Ceylon mapping tool.
Such a mapping mechanism has already been developed in the Eclipse Ceylon debugging support, and the code might be abstracted and factorized in the future, in order to also build a tool that gives an approximate Ceylon StackTrace (method calls) from the Java stackTrace (just as what we have in the Eclipse Debug view). Thus you wouldn't need adding fake approximate line number *only* for bug reporting.

@quintesse : it seems to me that the case when we might stop in a place without line numbers can be easily fixed in the IDE debugging support by:
- step-filtering such locations (locations without associated line numbers)
- (in case we still suspend in one of them) bringing the caret to the nearest Ceylon declaration containing the stackFrame location (using a specific `ISourceLocator` and the same Java-to-Ceylon mapping utility mentioned earlier).

So both requirements in favor of systematically adding line numbers for generated Java constructs that don't have a direct Ceylon equivalent, can be simply met by other means, either for stackTrace analysis or for the IDE debugging, mainly thanks to a Java-to-Ceylon mapping utility. 

Anyway, I'm not against adding as many line numbers as possible, as long as it doesn't bring problems such as those described in this issue (suspending on breakpoints in undesired or wrong locations).

Finally, I find @quintesse's proposal (about checking line number anomalies) interesting. This could be added as tests in the compiler, and would be easier to use than what we have for now and don't really use. However, we have to precisely define the rules about what is obviously or possibly wrong.

**Comment by tombentley** [comment]
I think you misunderstood me @davidfestal. I was meaning  when the code generator emits a bogus JCTree which is then rejected by javac. Right now the javac errors are reported with the Ceylon line number, so you know what the ceylon compiler transformed incorrectly. If we remove those line numbers to improve debugging it means bugs in the compiler will become more of a pain to work around and fix because it will be harder to figure out which Ceylon node is causing the problem.

**Comment by davidfestal** [comment]
Ah, OK, it's bugs during compile time ! So why not having a compiler option to enable disable line numbers that are added only for this need ?

**Comment by quintesse** [comment]
Let's first see what happens when we actually *fix* those numbers. I'm predicting a lot of the problems will already disappear or at least be less of a nuisance. Then we can look at the remaining problems to see how to solve them.

**Comment by davidfestal** [comment]
OK :)

**Comment by tombentley** [comment]
@davidfestal I've fixed the specific case you mention of the synthetic `$param$N` parameter being assigned to the local variable and being attributed to a Ceylon source line. 

I think you should open separate issues for other specific cases where you think we generate "bad" source line attribution.

I also just want to mention a perfectly good example of a "lines going backwards" case: Taking a reference:

    class Ref() {}
    void example() {
        value ref = Ref; // breakpoint on this line
        //...
        ref();
    }

I think with a breakpoint on the given line we should be breaking on the initial assignment, but also in each invocation. Thus we should attribute the local var decl and the code in the anonymous subclass of `AbstractCallable` to the same Ceylon source line. But maybe others disagree...?

**Comment by davidfestal** [comment]
In fact it didn't fix the main problem I opened this issue for. it seems that despite the line number info that you removed, there is still a line information that links line of the return with t he very beginning of the `$call$()` method (method codeIndex 0).

I think that the following Java lines [26](https://gist.github.com/davidfestal/f54616ba4b026063ed6a#file-doer-java-L26), [28](https://gist.github.com/davidfestal/f54616ba4b026063ed6a#file-doer-java-L28), [29](https://gist.github.com/davidfestal/f54616ba4b026063ed6a#file-doer-java-L29), [36](https://gist.github.com/davidfestal/f54616ba4b026063ed6a#file-doer-java-L36) should not be assigned to the Ceylon line [5](https://gist.github.com/davidfestal/f54616ba4b026063ed6a#file-doer-ceylon-L5)

IMO, Java lines [26](https://gist.github.com/davidfestal/f54616ba4b026063ed6a#file-doer-java-L26), [28](https://gist.github.com/davidfestal/f54616ba4b026063ed6a#file-doer-java-L28), [29](https://gist.github.com/davidfestal/f54616ba4b026063ed6a#file-doer-java-L29) should be assigned to the Ceylon line [3](https://gist.github.com/davidfestal/f54616ba4b026063ed6a#file-doer-ceylon-L3)

As for Java line [36](https://gist.github.com/davidfestal/f54616ba4b026063ed6a#file-doer-java-L36) it is also linked to line [5](https://gist.github.com/davidfestal/f54616ba4b026063ed6a#file-doer-ceylon-L5) probably because you retrieve the line previously associated to the arg$0$0 expression node, which (as I said before) should have been assigned to the Ceylon line [3](https://gist.github.com/davidfestal/f54616ba4b026063ed6a#file-doer-ceylon-L3).
But even then, it seems to me that you don't need to add the line of a variable you are just using, so that IMO Java line [36](https://gist.github.com/davidfestal/f54616ba4b026063ed6a#file-doer-java-L36) should only be associated to Ceylon line [2](https://gist.github.com/davidfestal/f54616ba4b026063ed6a#file-doer-ceylon-L2).

I reopen it, so that we can close it when the resulting bytecode doesn't associate the `$call$()` start codeIndex to Ceylon line 5 anymore

**Comment by quintesse** [comment]
One thing @davidfestal , I think we've seen from disassembling the Java .class files that we can't really trust in the output of your `JavaPositionsRetriever` because they don't coincide. There are far less positions being stored in the class file than what your utility shows.

**Comment by davidfestal** [comment]
Sure, but one of the positions I just mentioned generates an association with line 5 on the codeIndex 0 of the `$call()` method and this is the problem. I mentioned all this lines, because the `codeIndex0 <-> Ceylon line 5` link comes from one of them.
The positions retrieved  by the `JavaPositionsRetriever` can be trusted, but as as representative of the positions stored on the Javac AST nodes only. However these line numbers associated to AST nodes are used by the Javac bytecode generator to generate the  line number subset we can read inside the bytecode by `Javap`.

**Comment by quintesse** [comment]
Sure, so we can use to to see when the associated line number is obviously *wrong*, but we cannot say " it seems to me that you don't need to add the line of a variable you are just using, so that IMO Java line 36 should only be associated to Ceylon line 2." because you don't know if any of that information is actually written to the .class file or not.

**Comment by davidfestal** [comment]
Yes, you're right. I made a confusion with this sentence :-)

**Comment by tombentley** [comment]
@davidfestal, this issue mentions quite a few problems, and I'm trying to understand what your *biggest* problem is. You mention code index 0 being attributed to line 5 (and you want it attributed to line 3) of the source. This that the most important aspect of this issue?

**Comment by davidfestal** [comment]
Thanks @tombentley  ! The initial problem is fixed, and debug flow is really better. I'll check today that this didn't introduced any regression regarding #2047.

**Comment by davidfestal** [comment]
and afaict there was no regression regarding #2047.

**Comment by tombentley** [comment]
@davidfestal, I think it should be OK wrt #2047.

Right now I've only fixed up the line numbers for `Tree.FunctionArgument` and `Tree.MethodArgument`, not the other instantiations of `AbstractCallable`, so I guess I 'll need to do those too before I can really close this issue. But I'll wait for you to confirm you're happy with what we've got so far before I proceed.

**Comment by tombentley** [comment]
@davidfestal, I think it should be OK wrt #2047.

Right now I've only fixed up the line numbers for `Tree.FunctionArgument` and `Tree.MethodArgument`, not the other instantiations of `AbstractCallable`, so I guess I 'll need to do those too before I can really close this issue. But I'll wait for you to confirm you're happy with what we've got so far before I proceed.

**Comment by tombentley** [comment]
@davidfestal, I think it should be OK wrt #2047.

Right now I've only fixed up the line numbers for `Tree.FunctionArgument` and `Tree.MethodArgument`, not the other instantiations of `AbstractCallable`, so I guess I 'll need to do those too before I can really close this issue. But I'll wait for you to confirm you're happy with what we've got so far before I proceed.

**Comment by tombentley** [comment]
@davidfestal, I think it should be OK wrt #2047.

Right now I've only fixed up the line numbers for `Tree.FunctionArgument` and `Tree.MethodArgument`, not the other instantiations of `AbstractCallable`, so I guess I 'll need to do those too before I can really close this issue. But I'll wait for you to confirm you're happy with what we've got so far before I proceed.

**Comment by tombentley** [comment]
@davidfestal, I think it should be OK wrt #2047.

Right now I've only fixed up the line numbers for `Tree.FunctionArgument` and `Tree.MethodArgument`, not the other instantiations of `AbstractCallable`, so I guess I 'll need to do those too before I can really close this issue. But I'll wait for you to confirm you're happy with what we've got so far before I proceed.

**Comment by tombentley** [comment]
@davidfestal, I think it should be OK wrt #2047.

Right now I've only fixed up the line numbers for `Tree.FunctionArgument` and `Tree.MethodArgument`, not the other instantiations of `AbstractCallable`, so I guess I 'll need to do those too before I can really close this issue. But I'll wait for you to confirm you're happy with what we've got so far before I proceed.

**Comment by tombentley** [comment]
@davidfestal, I think it should be OK wrt #2047.

Right now I've only fixed up the line numbers for `Tree.FunctionArgument`
and `Tree.MethodArgument`, not the other instantiations of
`AbstractCallable`, so I guess I 'll need to do those too before I can
really close this issue. But I'll wait for you to confirm you're happy with
what we've got so far before I proceed.

**Comment by quintesse** [comment]
Once was enough Tom :D

**Comment by tombentley** [comment]
Haha! My connection is *terrible* today for some reason, and without the feedback of my message appearing I just kept hitting "Comment".

**Comment by tombentley** [comment]
In the same theme, if we have an method with multiple parameter lists:

```ceylon
Integer mpl
(Integer i)
(Integer j)
(Integer k)  
        => 
        i + j + k; 
```
do we expect to be able to break on the partial applications?

Likewise static member references:

```ceylon
Integer(Integer)(Integer) staticMethodReference 
        => 
        Integer.plus;
```

Note in this latter case, because there's no parameter lists we can't lay them out so we can apply a breakpoint to a particular partial application, but the question remains, should we be able to break when the first argument list is applied?

**Comment by gavinking** [comment]
@tombentley well it would be a nice-to-have, I guess, but partial application doesn't actually *enter* the function body, so not *necessary*.

**Comment by davidfestal** [comment]
@tombentley : I stumbled upon something else.

With the class `ZipFileVirtualFile` available [here](https://github.com/ceylon/ceylon-ide-common/blob/278132b3f0a3c345c398bad4b628c5a87f84b5c6/source/com/redhat/ceylon/ide/common/vfs/zip.ceylon#L129), it seems that the line [269](https://github.com/ceylon/ceylon-ide-common/blob/278132b3f0a3c345c398bad4b628c5a87f84b5c6/source/com/redhat/ceylon/ide/common/vfs/zip.ceylon#L269) is associated to several locations in the source code.

You can see it in the IntelliJ bytecode view of the generated code available [here](https://gist.github.com/davidfestal/582b243399d76008d306).
To summarize, the following methods have their first codeIndex wrongly associated with like 269 :
- the [ignored constructor](https://gist.github.com/davidfestal/582b243399d76008d306#file-gistfile1-txt-L79)
- the [`$getType$() method](https://gist.github.com/davidfestal/582b243399d76008d306#file-gistfile1-txt-L1030),
- the [synthetic bridge `findFile()` method](https://gist.github.com/davidfestal/582b243399d76008d306#file-gistfile1-txt-L1042),
- the [synthetic bridge `getParent()` method](https://gist.github.com/davidfestal/582b243399d76008d306#file-gistfile1-txt-L1056),
- the various synthetic `access$xxx()` methods.

So if we set a breakpoint on this line, instead of stopping only once, we stop many times in a very strange way compared to the expected control flow.  

I assume that there is a pending line number that is not reset in the generator, so that all these elements inherit from it.

**Comment by FroMage** [comment]
Good work Tom!

**Comment by FroMage** [comment]
Good work Tom!

**Comment by FroMage** [comment]
Good work Tom!

**Comment by FroMage** [comment]
Good work Tom!

**Comment by FroMage** [comment]
Good work Tom!

**Comment by FroMage** [comment]
Good work Tom!

**Comment by tombentley** [comment]
@davidfestal: Is that any better?

@FroMage:

```ceylon
for (i in 0:6) {
    print("Are you being sarcastic?");
}
```

**Comment by FroMage** [comment]
:)

**Comment by FroMage** [comment]
:)

**Comment by FroMage** [comment]
:)

**Comment by FroMage** [comment]
:)

**Comment by FroMage** [comment]
:)

**Comment by FroMage** [comment]
:)

**Comment by FroMage** [comment]
Nooooo…

**Comment by tombentley** [comment]
@davidfestal do you need any more changes for this?

**Comment by davidfestal** [comment]
I'm going to check very soon and let you know

**Comment by davidfestal** [comment]
@tombentley : I cannot step over / step into correctly inside the  function `initializeChildren()` ([here](https://github.com/ceylon/ceylon-ide-common/blob/278132b3f0a3c345c398bad4b628c5a87f84b5c6/source/com/redhat/ceylon/ide/common/vfs/zip.ceylon#L200)) anymore. It seems that lines inside the generated `ZipFileVirtualFile$1buildParents_` class contains code that is linked to the line [202](https://github.com/ceylon/ceylon-ide-common/blob/278132b3f0a3c345c398bad4b628c5a87f84b5c6/source/com/redhat/ceylon/ide/common/vfs/zip.ceylon#L202)). Sorry about that. I'll investigate more to undserstand what's happening.

**Comment by davidfestal** [comment]
@tombentley : I'm still trying to fight against a bug that prevents me to check the whole, but I can already say 2 things :
- Thanks for removing the wrong line on `$getType$()`.
- However, since this change, the first codeIndex of synthetic methods `findFile()`, `getParent()` and `access$xxx()` are all now associated to the line [129](https://github.com/ceylon/ceylon-ide-common/blob/278132b3f0a3c345c398bad4b628c5a87f84b5c6/source/com/redhat/ceylon/ide/common/vfs/zip.ceylon#L129).

This is strange, because, for example in the case of the synthetic `access$xxx()` methods, the delegate method is each time correctly assigned to the corresponding source line : 

- `access$200()` (wrongly associated to line 129) delegates to `getZipFile$priv()` (correctly associated to line 130)
- `access$300()` (wrongly associated to line 129) delegates to `initializeChildren$priv$()` (correctly associated to line 201).

Since the line associated to these synthetic methods changed with your last change from line 269 to line 129, I assume they are still inheriting from some pending `at()` instruction in the transformation code.
It seems that these synthetic methods should either have to line number associated, or have the same line number as the delegate method.

**Comment by tombentley** [comment]
With regard to the `access$NNN()` methods, those are generated by `javac` (during `Lower`, I think) to cope with things like inner classes which capture `private` methods of their outer class. `javac` seems to attribute the invocation of the captured `private` method within the `access$NNN()` method to the line number of the `JTClassDecl`, rather than the declaration of the delegated-to method. Note that the JDT compiler seems to attribute them to the delegated-to method. Anyway, I'm not sure there's anything I can do about those.

I'll need to look into the cases of `findFile()` and `getParent()` to understand what's going on there.

**Comment by tombentley** [comment]
The `getParent()` (and I assume `findFile()`) methods you're stepping into are bridge methods, again generated by javac, this time to cope with covariant return types. Again, I'm not sure there's very much I can easily do to change the line numbers those are attributed to. At the point the access methods and bridge methods are generated by Javac it basically only has a `MethodSymbol` for the thing it's delegating to, and those symbols don't contain line number information.

**Comment by FroMage** [comment]
Aren't those methods marked as `synthetic` though?

**Comment by gavinking** [comment]
I'm surprised the debugger would stop at a synthetic method...

**Comment by tombentley** [comment]
Yes, they're all `synthetic` according to `javap`. I note that @davidfestal doesn't *actually* say he's stepping into them though, he's just objecting to their `LineNumberTable`s.

**Comment by davidfestal** [comment]
@gavinking : if we don't activate step filers, we will stop in them yes. it's JDT default behavior.
@tombentley : I knew that `access$xxx()` were synthetic methods delegating to private methods of the inner class, and that the `getParent()` and `findFile()` were bridge methods, all generated by Javac.
However, before your commit ceylon/ceylon-compiler@fdac993d8dc9ed4113f1dc050f8cbc827e347125, those synthetic methods were assigned the 269 line number, while after this commit, they received the 129 line number. This is why I assumed we might be able to somehow change (maybe indirectly) their associated line number.

**Comment by tombentley** [comment]
@davidfestal well I suppose that commit proves that we can change them, but only *en-masse*, not individually.

**Comment by davidfestal** [comment]
I see. It is probably good like that then.

**Comment by tombentley** [comment]
@davidfestal apart from the work on ceylon/ceylon-ide-eclipse#1260, so there anything this issue requires?

**Comment by davidfestal** [comment]
No I think it's OK... at least sufficient to have a pretty good debugging experience IMO. Thanks @tombentley !

**Comment by tombentley** [comment]
OK then, closing.
